### PR TITLE
GH-409: Short syntax for React fragment

### PR DESCRIPTION
**Pull request by qtran-n4**
Task: https://github.com/eclipse/n4js/issues/409

### CODE REVIEW

**Comment by qtran-n4** [comment]
Builds are green!

File: plugins/org.eclipse.n4js.model/model/N4JS.xcore [hunk 1]
```diff
@@ -2287,6 +2287,9 @@ class JSXElement extends Expression, JSXChild {
 	contains JSXElementName jsxClosingName
 }
 
+class JSXFragment extends Expression, JSXChild {
```
Comment (mmews-n4) [inline plugins/org.eclipse.n4js.model/model/N4JS.xcore]: Can the class `JSXElement` **extend** this new class `JSXFragment`? In case it can: It would be nice to introduce a new private method in `CFEChildren` to reduce copied code.

File: plugins/org.eclipse.n4js.transpiler.es/src/org/eclipse/n4js/transpiler/es/transform/JSXTransformation.xtend [hunk 2]
```diff
@@ -91,15 +95,21 @@ class JSXTransformation extends Transformation {
 		if(!inJSX) {
 			return; // this transformation is not applicable
 		}
+
+		// Transform JSXFragments and JSXElements	
+		val jsxFragments = collectNodes(state.im, JSXFragment, true);
 		val jsxElements = collectNodes(state.im, JSXElement, true);
-		if(jsxElements.empty) {
-			return; // nothing to transform
+		if (jsxFragments.isEmpty && jsxElements.isEmpty) {
```
Comment (mmews-n4) [inline plugins/org.eclipse.n4js.transpiler.es/src/org/eclipse/n4js/transpiler/es/transform/JSXTransformation.xtend]: Also here: In case of a common superclass `JSXFragment` refactor if possible to reduce code duplication.

File: plugins/org.eclipse.n4js.transpiler.es/src/org/eclipse/n4js/transpiler/es/transform/JSXTransformation.xtend [hunk 3]
```diff
@@ -131,6 +141,14 @@ class JSXTransformation extends Transformation {
 		}
 		return getSymbolTableEntryOriginal(elementFactoryFunction, true);
 	}
+	
+	def private SymbolTableEntryOriginal prepareFragmentComponent() {
```
Comment (mmews-n4) [inline plugins/org.eclipse.n4js.transpiler.es/src/org/eclipse/n4js/transpiler/es/transform/JSXTransformation.xtend]: Maybe this and the above methods can share same code?

File: plugins/org.eclipse.n4js.transpiler.es/src/org/eclipse/n4js/transpiler/es/transform/JSXTransformation.xtend [hunk 4]
```diff
@@ -142,6 +160,16 @@ class JSXTransformation extends Transformation {
 		replace(elem, convertJSXElement(elem));
 	}
 
+	def private void transformJSXFragment(JSXFragment elem) {
```
Comment (mmews-n4) [inline plugins/org.eclipse.n4js.transpiler.es/src/org/eclipse/n4js/transpiler/es/transform/JSXTransformation.xtend]: Code duplication?

File: plugins/org.eclipse.n4js.transpiler.es/src/org/eclipse/n4js/transpiler/es/transform/JSXTransformation.xtend [hunk 5]
```diff
@@ -155,6 +183,22 @@ class JSXTransformation extends Transformation {
 		);
 	}
 
+	def private ParameterizedCallExpression convertJSXFragment(JSXFragment elem) {
+//		var identifiableElem = TypesFactory.eINSTANCE.createIdentifiableElement();
+//		identifiableElem.name = "React.Fragment";
+		var idRef = ImFactory.eINSTANCE.createIdentifierRef_IM();
+//		idRef.id = identifiableElem;
```
Comment (mmews-n4) [inline plugins/org.eclipse.n4js.transpiler.es/src/org/eclipse/n4js/transpiler/es/transform/JSXTransformation.xtend]: Are those four lines above necessary?

File: plugins/org.eclipse.n4js.transpiler.es/src/org/eclipse/n4js/transpiler/es/transform/JSXTransformation.xtend [hunk 6]
```diff
@@ -155,6 +183,22 @@ class JSXTransformation extends Transformation {
 		);
 	}
 
+	def private ParameterizedCallExpression convertJSXFragment(JSXFragment elem) {
```
Comment (mmews-n4) [inline plugins/org.eclipse.n4js.transpiler.es/src/org/eclipse/n4js/transpiler/es/transform/JSXTransformation.xtend]: Code duplication?

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/n4jsx/ReactHelper.xtend [hunk 7]
```diff
@@ -205,7 +220,18 @@ class ReactHelper {
...
+	def private TClass lookUpReactFragmentComponent(TModule module) {
+		if (module !== null) {
+			for (Type currTopLevelType : module.getTopLevelTypes()) {
+				if (currTopLevelType instanceof TClass
+						&& REACT_FRAGMENT_NAME.equals(currTopLevelType.getName())) {
+					return currTopLevelType as TClass;
+				}
+			}
+		}
+		return null;
```
Comment (mmews-n4) [inline plugins/org.eclipse.n4js/src/org/eclipse/n4js/n4jsx/ReactHelper.xtend]: Not sure, but: In case this method returns null, would there be a validation to handle that?

File: tests/org.eclipse.n4js.n4jsx.spec.ui.tests/xpectTests/compile/replacer.n4js [hunk 8]
```diff
@@ -13,12 +13,7 @@
 // See: https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify
 export public function replacer(key: string, value: any): any {
 	if (value instanceof Function) {
-		//Extract the function name out of the function string
-		//Note: this does not work for anonymous functions! But this is just a test.
-		let fnName = (value as Function).toString();
-		fnName = fnName.substring('function '.length);
-		fnName = fnName.substring(0, fnName.indexOf('('));
-		return fnName.trim()? fnName.trim() : "function" ;
+		return "function";
```
Comment (mmews-n4) [inline tests/org.eclipse.n4js.n4jsx.spec.ui.tests/xpectTests/compile/replacer.n4js]: Why did you change this implementation?

File: plugins/org.eclipse.n4js.transpiler.es/src/org/eclipse/n4js/transpiler/es/transform/JSXTransformation.xtend (hunk 5 above)
Comment (mor-n4) [inline plugins/org.eclipse.n4js.transpiler.es/src/org/eclipse/n4js/transpiler/es/transform/JSXTransformation.xtend]: no need for these comment, I guess?

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/validation/validators/N4JSXValidator.xtend [hunk 9]
```diff
@@ -147,6 +147,9 @@ class N4JSXValidator extends AbstractN4JSDeclarativeValidator {
 	 */
 	@Check
 	def public void checkReactElementBinding(JSXElement jsxElem) {
+		if (jsxElem.jsxElementName === null) {
+			return;
+		}
```
Comment (mor-n4) [inline plugins/org.eclipse.n4js/src/org/eclipse/n4js/validation/validators/N4JSXValidator.xtend]: why is this required, now?

File: tests/org.eclipse.n4js.n4jsx.spec.ui.tests/xpectTests/compile/replacer.n4js (hunk 8 above)
Comment (qtran-n4) [reply tests/org.eclipse.n4js.n4jsx.spec.ui.tests/xpectTests/compile/replacer.n4js]: The original function tries to extract the function name but it does not work correctly. Also for the test purpose, `function` is all what we need.

File: plugins/org.eclipse.n4js.transpiler.es/src/org/eclipse/n4js/transpiler/es/transform/JSXTransformation.xtend (hunk 5 above)
Comment (qtran-n4) [reply plugins/org.eclipse.n4js.transpiler.es/src/org/eclipse/n4js/transpiler/es/transform/JSXTransformation.xtend]: Removed. Done.

Comment (qtran-n4) [reply plugins/org.eclipse.n4js.transpiler.es/src/org/eclipse/n4js/transpiler/es/transform/JSXTransformation.xtend]: Removed. Done.

File: plugins/org.eclipse.n4js.model/model/N4JS.xcore (hunk 1 above)
Comment (qtran-n4) [reply plugins/org.eclipse.n4js.model/model/N4JS.xcore]: I am afraid not since conceptually JSXElement is not a special case of JSXFragment...

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/validation/validators/N4JSXValidator.xtend (hunk 9 above)
Comment (qtran-n4) [reply plugins/org.eclipse.n4js/src/org/eclipse/n4js/validation/validators/N4JSXValidator.xtend]: Removed. Relicts of attempt to make jsxElementName optional...

File: plugins/org.eclipse.n4js.model/model/N4JS.xcore (hunk 1 above)
Comment (qtran-n4) [reply plugins/org.eclipse.n4js.model/model/N4JS.xcore]: Done. Introduce JSXAbstractElement as basic class for JSXFragment and JSXElement

**Comment by qtran-n4** [comment]
@mmews-n4 
(2) Try this in an empty file.n4jsx: let x = </>; Should create some problem on the command line... maybe you can do something about that?

=> What do you mean?  That is correct N4JS code...

**Comment by mmews-n4** [comment]
@qtran-n4 I know. So, if I try it out in my IDE, I get the following problem:
`Severe error occurred while transpiling module MyModule. Check error log for details about the failure.`

**Comment by qtran-n4** [comment]
@mmews-n4 : Nice catch! Fixed in N4JSXValidator

**Comment by qtran-n4** [comment]
Thanks for the reviews!
